SPM-1705: Refactor vmaas-sync recalc event

michalslomczynski opened the pull request:
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

michalslomczynski commented:
Regarding refactor:
- Not a strong preference against previous variable names
- Very strong preference to get rid of this `groupedData` struct and `data` variable names.

psegedy commented:
I haven't had time to look at it properly. Wouldn't it be possible to completely remove everything related to `accountInventories` and use only `EvalDataSlice` struct?

psegedy commented on base/mqueue/platform_event.go:
```diff
@@ -83,46 +85,51 @@ func batchSize(grouped groupedData) int {
 }
 
-func (data *InventoryAIDs) groupData() (int, *groupedData) {
+func (accs *InventoryAccounts) getAccountInventories() (int, *accountInventories, *orgIDs) {
```
there was an intention of having the same API/function name for multiple receiver types, no strong preference tho

michalslomczynski replied:
I see the reason for `WriteEvents` same function name, it's part of `MessageData` interface, but I have no clue for these getters. What is the reason?

It's better to use names that suggest what variable holds. These new names may be improved, but `InventoryAccounts` is still lesser set than `groupedData`, that can refer to almost anything :smile:

psegedy commented on base/mqueue/platform_event.go:
```diff
@@ -131,21 +138,14 @@ func (data *InventoryAIDs) WriteEvents(ctx context.Context, w Writer) error {
-		groupedReqIDVal = *groupedReqID
-	}
+func (evals *EvalDataSlice) WriteEvents(ctx context.Context, w Writer) error {
+	// invs, reqs, orgIDs never nil
```
what if someone will return nil in `getAccountEvalData` in the future? It's still a valid return value.
Can't we make `reqs=nil` and use EvalDataSlice instead of the InventoryAccounts? `reqs`'s the only difference in those 2 `WriteEvents` functions. Just a thought

michalslomczynski replied:
I refactored this because it was mixed approach of assuming dereferencing is one time safe and other is not, while they all are the same, never nil pointers, so lack of uniformity.

Now it's better because if you handle nil pointer inside getter function, you don't repeat itself every time you use it and do not obfuscate other functions. It's very popular to handle nil pointers in getters and return default value if it's nil - empty struct in this case.

If one changes getter, he ought to check it's receivers.

Mayve EvalDataSlice can be reused indeed. I'll take a look.

michalslomczynski replied:
Slice in Go is just
```
type SliceHeader struct {
	Data [uintptr](https://pkg.go.dev/builtin#uintptr)
	Len  [int](https://pkg.go.dev/builtin#int)
	Cap  [int](https://pkg.go.dev/builtin#int)
}
```
https://pkg.go.dev/reflect#SliceHeader
So all the heavy stuff is pointer anyway here, so probably it will be faster to return from these getters as value, and we will get rid of nil pointer issue :slightly_smiling_face:

michalslomczynski replied:
It's the same performance
```
By value
    1672            646303 ns/op
By reference
    1602            658101 ns/op
```
Let's just drop pointers from getters.

psegedy replied:
I agree with passing by values, that's a good point. Please try to look if it is possible to use only `EvalDataSlice`

psegedy commented on tasks/vmaas_sync/repo_based.go:
```diff
@@ -38,18 +38,19 @@ func getCurrentRepoBasedInventoryIDs() ([]mqueue.InventoryAID, error) {
 		Where("repo.name IN (?)", repos).
 		Order("sp.rh_account_id").
-		Select("distinct sp.inventory_id, sp.rh_account_id").
+		Select("distinct sp.inventory_id, sp.rh_account_id, ra.org_id").
```
I see it was there before, I just wonder why the `distinct` is needed

michalslomczynski replied:
Let's ping pepa :smile:

michalslomczynski replied:
Mim authored this one, he will know. Looks like safety check or corner case handle. Do you think it can be a problem?

psegedy replied:
not a problem, just curious. Distinct might be slowing this down a bit

michalslomczynski replied:
Better call @MichaelMraka.

michalslomczynski replied:
@psegedy This should be fine.

michalslomczynski commented:
@psegedy I replaced it in last commit, but removed `EvalData`.

I personally prefer `InventoryAccount` over `EvalData` since you do not know what `EvalData` is without checking it's implementation. It is not improved readability if you have to do this IMO. `InventoryAccount` name is more specific. This struct just holds the account of the inventory, and the name suggest exactly that, simple.

MichaelMraka commented on base/database/testing.go:
```diff
@@ -442,3 +442,8 @@ func CheckBaselineDeleted(t *testing.T, baselineID int) {
+
+func GetAllSystems(t *testing.T) (systems []*models.SystemPlatform) {
+	assert.Nil(t, Db.Model(&models.SystemPlatform{}).Select("*").Order("rh_account_id").Scan(&systems).Error)
+	return systems
```
Select() is redundant, `Db.Model(&models.SystemPlatform{}).Order("rh_account_id").Scan(&systems).Error` is enough.

MichaelMraka commented:
@michalslomczynski, I have hard times to understand this PR. Can you please split it into
- necessary orgId changes,
- independent refactoring PR?

michalslomczynski commented:
Necessary OrgID changes:
[SPM-1705: Add OrgID field to eval recalc messages](https://github.com/RedHatInsights/patchman-engine/pull/1061/commits/7e3101f70d7dd71b82b3ed349f01a640da4c9d78)

michalslomczynski commented:
To sum up refactor in case it's hard to track what was changed:
- Removed improper use of Go types
- Removed improper use of pointers
~~- Renamed inventoryAID/inventoryAIDs -> InventoryAccount/InventoryAccounts~~ Reverted this in favor of `EvalDataSlice` Patrick suggested
- Removed one of `MessageData` interface implementation because it was redundant.

michalslomczynski commented:
@psegedy I am fine with EvalDataSlice and EvalData names, but we have this weird old variables all over the place, eg. `inventoryAIDs = EvalDataSlice{...}`.

psegedy commented:
@michalslomczynski not a big deal IMO and it would require renaming also some functions. Maybe something like `inventoryAccounts` would be more appropriate now. It used to contain only inventoryID and accountID, but now it also has orgID. `inventoryAIDs` might not be very descriptive, I guess it should stand for `inventory account ids` which is technically still true with orgIDs, I'd let it be as it is.
